src/halmos/calldata.py

```diff
@@ -210,3 +247,71 @@ def head_size(x):
         static = len(tails) == 0
 
         return EncodingResult(heads + tails, total_size, static)
+
+
+@dataclass(frozen=True)
+class FunctionInfo:
+    name: str | None = None
+    sig: str | None = None
+    selector: str | None = None
+
+
+def str_abi(item: dict) -> str:
+    def str_tuple(args: list) -> str:
+        ret = []
+        for arg in args:
+            typ = arg["type"]
+            match = re.search(r"^tuple((\[[0-9]*\])*)$", typ)
+            if match:
+                ret.append(str_tuple(arg["components"]) + match.group(1))
+            else:
+                ret.append(typ)
+        return "(" + ",".join(ret) + ")"
+
+    if item["type"] != "function":
+        raise ValueError(item)
+    return item["name"] + str_tuple(item["inputs"])
+
+
+def find_abi(abi: list, fun_info: FunctionInfo) -> dict:
+    funname, funsig = fun_info.name, fun_info.sig
+    for item in abi:
+        if (
+            item["type"] == "function"
+            and item["name"] == funname
+            and str_abi(item) == funsig
+        ):
+            return item
+    raise ValueError(f"No {funsig} found in {abi}")
+
+
+def mk_calldata(
+    abi: list,
+    fun_info: FunctionInfo,
+    cd: ByteVec,
+    dyn_param_size: DynamicParams,
+    args: HalmosConfig,
+    arrlen: dict[str, int] = None,
+    new_symbol_id: Callable = None,
+) -> None:
```

Comment:
would it be possible to have `mk_calldata` return `cd` instead of mutating an input parameter?

Reply:
will be done in a separate pr. added a todo comment for that.

Reply:
done